### PR TITLE
batch fetching status lists

### LINKED ISSUE #27: Add suggestions for API for batch fetching statuslists
If a verifier wants all statuslists for a specific type of Referenced Token, e.g. for offlien caching, then it would be handy, if the issuer provides a API endpoint to get those in one call, e.g. as an array.

### PR DESCRIPTION
**Pull request by c2bo**
closes #27

Rendered Version: https://drafts.oauth.net/draft-ietf-oauth-status-list/c2bo/batch-fetching/draft-ietf-oauth-status-list.html

### CODE REVIEW

File: draft-ietf-oauth-status-list.md [hunk 1]
```diff
@@ -219,6 +220,7 @@ This section defines the structure for a CBOR-encoded Status List:
 * The `StatusList` structure is a map (Major Type 5) and defines the following entries:
   * `bits`: REQUIRED. Unsigned int (Major Type 0) that contains the number of bits per Referenced Token in the Status List. The allowed values for `bits` are 1, 2, 4 and 8.
   * `lst`: REQUIRED. Byte string (Major Type 2) that contains the Status List as specified in [](#status-list-json).
+  * `aggregation_uri`: OPTIONAL. Byte string (Major Type2) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:223]: Why are we repeating the definitions for CBOR Status List? In the CWT Status List Token we just reference to the JWT.Should we do the same here?

File: draft-ietf-oauth-status-list.md [hunk 2]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
 
 TBD
 
+# Batch Fetching {#batch-fetching}
+
+To allow a Relying Party to fetch all status lists for a specific type of credential or a specific issuer, an optional mechanism is provided to retrieve an Object containing a list of Status Lists called Status List Aggregation. This mechanism is intended to optimize fetching and caching mechanisms and allow offline validation of the status.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:469]:
```suggestion
To allow a Relying Party to fetch all Status Lists for a specific Referenced Token or issuer, an optional mechanism is provided to retrieve a list of URIs to all Status List Tokens called Status List Aggregation. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status.
```

File: draft-ietf-oauth-status-list.md [hunk 3]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+
+There are two options that a Relying Party can use to discover the relevant Status List Aggregation URI.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:471]:
```suggestion
There are two options for a Relying Party to retrieve the Status List Aggregation.
```

File: draft-ietf-oauth-status-list.md [hunk 4]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+An issuer MAY support any of these mechanisms:
+
+- Issuer metadata: The issuer publishes an URI which links to Status List Aggregation
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:474]:
```suggestion
- Issuer metadata: The issuer of the Referenced Token publishes an URI which links to Status List Aggregation, e.g. in publicly available metadata of an issuance protocol
```

File: draft-ietf-oauth-status-list.md [hunk 5]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+- Status List Parameter: The issuer can choose to add an additional claim to the Status List (Token) that links to the Status List Aggregation.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:475]:
```suggestion
- Status List Parameter: The issuer of the Referenced Token includes an additional claim to the Status List (Token) that links to the Status List Aggregation.
```

File: draft-ietf-oauth-status-list.md [hunk 6]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:476]: maybe turn these chapters around? Now that we mentioned the two mechanisms, giving the details on those fits better

File: draft-ietf-oauth-status-list.md [hunk 7]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+## Status List Aggregation in JSON Format
+
+This section defines the structure for a JSON-encoded Status List Aggregation:
+
+* `status_lists`: REQUIRED. Contains an array of strings, each containing an URI to a Status List (Token)
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:481]:
```suggestion
* `status_lists`: REQUIRED. The `status_lists` claim MUST be an array of strings, each containing an URI to a Status List (Token).
```

File: draft-ietf-oauth-status-list.md [hunk 8]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+
+The Status List Aggregation URI provides a list of Status List URIs. This aggregation in JSON and the media type return SHOULD be `application/json`.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:483]: maybe explain the fetching logic and move the the media type to the Section 8?

File: draft-ietf-oauth-status-list.md [hunk 9]
```diff
@@ -462,6 +464,50 @@ If caching is required (e.g., to enable the use of alternative mechanisms for ho
...
+
+The issuer MAY link to the Status List Aggregation URI in metadata that can be provided by different means like .well-known metadata as is used in OAuth and OpenID, or via a VICAL extension for ISO mDoc / mDL.
+
+> OpenID4VCI Example
+
+> VICAL example
+
+## Status List Parameter
+
+The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party can choose to follow this link and retrieve an up-to-date list of the relevant Status Lists.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:509]:
```suggestion
The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party may use this link to retrieve an up-to-date list of relevant Status Lists.
```

**Review by paulbastian — COMMENTED** [review]
Looks already solid. Let's move the fetching and media type to Section 8

File: draft-ietf-oauth-status-list.md (hunk 1 above)
Comment (c2bo) [reply draft-ietf-oauth-status-list.md:223]: Hmm, we need to define the major type here, but yeah we can probably make this a bit cleaner - let's briefly discuss tomorrow

Comment (tplooker) [reply draft-ietf-oauth-status-list.md:223]: I think we should use major type 3 - text string.

File: draft-ietf-oauth-status-list.md [hunk 10]
```diff
@@ -205,6 +205,7 @@ This section defines the structure for a JSON-encoded Status List:
 * `status_list`: REQUIRED. JSON Object that contains a Status List. The object contains exactly two claims:
    * `bits`: REQUIRED. JSON Integer specifying the number of bits per Referenced Token in the Status List (`lst`). The allowed values for `bits` are 1,2,4 and 8.
    * `lst`: REQUIRED. JSON String that contains the status values for all the Referenced Tokens it conveys statuses for. The value MUST be the base64url-encoded (as defined in Section 2 of {{RFC7515}}) Status List as specified in [](#status-list).
+   * `aggregation_uri`: OPTIONAL. An URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:208]:
```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md (hunk 1 above)
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:223]:
```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type3) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md [hunk 11]
```diff
@@ -461,6 +463,47 @@ The HTTP response SHOULD use gzip Content-Encoding as defined in {{RFC9110}}.
...
+
+## Status List Parameter
+
+The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party may use this link to retrieve an up-to-date list of relevant Status Lists.
+
+## Status List Aggregation in JSON Format
+
+This section defines the structure for a JSON-encoded Status List Aggregation:
+
+* `status_lists`: REQUIRED. The `status_lists` claim MUST be an array of strings, each containing an URI to a Status List (Token).
```
Comment (paulbastian) [inline draft-ietf-oauth-status-list.md:490]:
```suggestion
* `status_lists`: REQUIRED. JSON array of strings that contains URIs linking to Status List (Tokens).
```

**Review by paulbastian — CHANGES_REQUESTED** [review]

File: draft-ietf-oauth-status-list.md [hunk 12]
```diff
@@ -205,6 +205,7 @@ This section defines the structure for a JSON-encoded Status List:
 * `status_list`: REQUIRED. JSON Object that contains a Status List. The object contains exactly two claims:
    * `bits`: REQUIRED. JSON Integer specifying the number of bits per Referenced Token in the Status List (`lst`). The allowed values for `bits` are 1,2,4 and 8.
    * `lst`: REQUIRED. JSON String that contains the status values for all the Referenced Tokens it conveys statuses for. The value MUST be the base64url-encoded (as defined in Section 2 of {{RFC7515}}) Status List as specified in [](#status-list).
+   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```
[the remark this answers is not in the document]
Comment (c2bo) [reply draft-ietf-oauth-status-list.md:208]:
```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md [hunk 13]
```diff
@@ -219,6 +220,7 @@ This section defines the structure for a CBOR-encoded Status List:
 * The `StatusList` structure is a map (Major Type 5) and defines the following entries:
   * `bits`: REQUIRED. Unsigned int (Major Type 0) that contains the number of bits per Referenced Token in the Status List. The allowed values for `bits` are 1, 2, 4 and 8.
   * `lst`: REQUIRED. Byte string (Major Type 2) that contains the Status List as specified in [](#status-list-json).
+  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```
[the remark this answers is not in the document]
Comment (c2bo) [reply draft-ietf-oauth-status-list.md:223]:
```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md [hunk 14]
```diff
@@ -461,6 +463,47 @@ The HTTP response SHOULD use gzip Content-Encoding as defined in {{RFC9110}}.
 
 TBD
 
+# Batch Fetching {#batch-fetching}
```
[the remark this answers is not in the document]
Comment (c2bo) [reply draft-ietf-oauth-status-list.md:466]:
```suggestion
# Status List Aggregation {#batch-fetching}
```

**Comment by c2bo** [comment]
I'd merge this as is and move things to section 8 with another PR that perhaps also finds a better name than `aggregation_uri` if that's fine w/ you @paulbastian @tplooker?

**Review by paulbastian — APPROVED** [review]

File: draft-ietf-oauth-status-list.md [hunk 15]
```diff
@@ -206,6 +206,7 @@ This section defines the structure for a JSON-encoded Status List:
 * `status_list`: REQUIRED. JSON Object that contains a Status List. It MUST contain at least the following claims:
    * `bits`: REQUIRED. JSON Integer specifying the number of bits per Referenced Token in the Status List (`lst`). The allowed values for `bits` are 1,2,4 and 8.
    * `lst`: REQUIRED. JSON String that contains the status values for all the Referenced Tokens it conveys statuses for. The value MUST be the base64url-encoded (as defined in Section 2 of {{RFC7515}}) Status List as specified in [](#status-list).
+   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```
Comment (tplooker) [inline draft-ietf-oauth-status-list.md:209]:
```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains a URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md [hunk 16]
```diff
@@ -220,6 +221,7 @@ This section defines the structure for a CBOR-encoded Status List:
 * The `StatusList` structure is a map (Major Type 5) and defines the following entries:
   * `bits`: REQUIRED. Unsigned int (Major Type 0) that contains the number of bits per Referenced Token in the Status List. The allowed values for `bits` are 1, 2, 4 and 8.
   * `lst`: REQUIRED. Byte string (Major Type 2) that contains the Status List as specified in [](#status-list-json).
+  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```
Comment (tplooker) [inline draft-ietf-oauth-status-list.md:224]:
```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains a URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

File: draft-ietf-oauth-status-list.md [hunk 17]
```diff
@@ -449,6 +451,47 @@ The HTTP response SHOULD use gzip Content-Encoding as defined in {{RFC9110}}.
 
 TBD
 
+# Status List Aggregation {#batch-fetching}
+
+Status List Aggregation is an optional mechanism to retrieve a list of URIs to all Status List Tokens, allowing a Relying Party to fetch all relevant Status Lists for a specific type of Referenced Token or issuer. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status.
```
Comment (tplooker) [inline draft-ietf-oauth-status-list.md:456]:
```suggestion
Status List Aggregation is an optional mechanism to retrieve a list of URIs to all Status List Tokens, allowing a Relying Party to fetch all relevant Status Lists for a specific type of Referenced Token or issuer. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status of a reference token.
```

**Review by tplooker — APPROVED** [review]
LGTM as a first cut with the minor language changes I highlighted.

**Pull request merged**